### ISSUE TITLE
Run exe or shell command without cmd window popup

### ISSUE THREAD

**Issue by harg** [federico-terzi/espanso#249]
Thank you for this great software!
On windows, is it possible to run an exe or a shell command without the flashing cmd window?

**Comment by federico-terzi** [comment]
Hey @harg,

Thank you for pointing this out, I'll investigate and let you know if I find something. Hopefully we can get this shipped in the next release!

Cheers :)

**Comment by harg** [comment]
I do not know Rust but you can do it like that in Go : https://play.golang.org/p/2LAEm2LQc-X

This is the win32 api : ShowWindow(... SW_HIDE)
https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-showwindow

**Comment by federico-terzi** [comment]
Thank you @harg, this is definitely very useful starting point to solve the problem

**Comment by federico-terzi** [comment]
I may have solved this by changing the default shell from CMD to Powershell on Windows (as CMD had also other problems). Would you have any problem with using Powershell?

**Comment by harg** [comment]
It's fine for me but I think some users might want to keep cmd...
Is it possible to have both and select the desired shell in the config file? (powershell by default?)

**Comment by federico-terzi** [comment]
> It's fine for me but I think some users might want to keep cmd...
> Is it possible to have both and select the desired shell in the config file? (powershell by default?)

Fair point, the best solution then is to add another param to select the desired shell.
What do you think?

**Comment by harg** [comment]
Yep!
There is a third option on Windows : WSL (Windows Subsystem for Linux) but you invoke it via cmd or powershell. 

So if you can do somehing like this  :  `Get-ChildItem | wsl grep -i "document"` 
We are good to go ! :)

**Comment by federico-terzi** [comment]
This is a wonderful idea! I use the WSL all the time but I never thought about integrating it with espanso. That could improve its capability a lot on Windows.

**Comment by federico-terzi** [comment]
I made a few tests, and unfortunately, the behavior still occurs. 
Digging a bit more into the problem it seems like indeed a solution is possible, but is not trivial.
I also tried with the approach suggested in your Go example, but unfortunately even that does flash the CMD window for a short while.

A possible solution would be to use the [rust-subprocess](https://github.com/hniksic/rust-subprocess) library, hopefully there is an option to specify the required flag ([I've opened an issue for that](https://github.com/hniksic/rust-subprocess/issues/40)).

**Comment by federico-terzi** [comment]
As a note for myself, this could also be another solution: https://stackoverflow.com/questions/59692146/is-it-possible-to-use-the-standard-library-to-spawn-a-process-without-showing-th

**Comment by BNolet** [comment]
Happy to see this issue is still being worked on or at least an intention :) Was so happy to find a cross-platform solution but the powershell window is a pretty big deal-breaker at least for now

**Comment by federico-terzi** [comment]
I made some more experimentation and I might have found the correct approach, so I'll try to ship it in the next release. For reference:

We first need to activate the `windows_subsystem` to avoid associating a console to espanso, by placing this line as the topmost command in the main file:

```
#![windows_subsystem = "windows"]
```

Then, when spawning the powershell/cmd command, use the `CONSOLE_NO_WINDOW` flag with:

```
use std::os::windows::process::CommandExt;

Command::new("powershell").arg("-Command")
        .creation_flags(0x08000000)
...
```

I already tried this flag before but it didn't work as expected. The answer is that without the first `windows_subsystem` declaration, that flag has no effect.

As a result though, the console output doesn't work anymore (which is useful when developing), so to solve that problem we have to attach to the parent console with:

```
unsafe {winapi::um::wincon::AttachConsole(0xFFFFFFFF)};
```

**Comment by BNolet** [comment]
wow damn, sounds like a lot of moving parts! Thanks for the update on this one :)

**Comment by federico-terzi** [comment]
Just wanted to say, this bug will be finally be solved in the upcoming release!
